### PR TITLE
[IOTDB-2764] Refine the consensus layer framework and add examples

### PR DESCRIPTION
**Pull request by OneSizeFitsQuorum**
- Improve the example of using consensus layer
- Define `ByteBufferConsensusRequest` to allow the deserialization pushed down into the state machine, which is beneficial to avoid forwarding several rounds of multiple deserialization overhead.
- Abstract `BaseStateMachine` to deserialize `IConsenesusRequest` so that its subclasses can focus only on standalone logic.

### CODE REVIEW

**Review by JackieTien97 — APPROVED** [review]

**Review by wangchao316 — APPROVED** [review]
LGTM

**Review by CRZbulabula — APPROVED** [review]

File: server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java [hunk 1]
```diff
@@ -39,16 +38,14 @@ public void start() {}
   public void stop() {}
 
   @Override
-  public TSStatus write(IConsensusRequest request) {
-    if (request instanceof InsertRowPlan) {
-      logger.info("Execute write plan : {}", request);
-    }
-    return new TSStatus(200);
+  public TSStatus write(PhysicalPlan plan) {
+    logger.info("Execute write plan in DataRegionStateMachine : {}", plan);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java:42]:
```suggestion
    logger.info("Execute write plan in SchemaRegionStateMachine : {}", plan);
```

File: server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java [hunk 2]
```diff
@@ -17,20 +17,19 @@
...
-import org.apache.iotdb.db.qp.physical.crud.InsertRowPlan;
+import org.apache.iotdb.db.qp.physical.PhysicalPlan;
+import org.apache.iotdb.rpc.TSStatusCode;
 import org.apache.iotdb.service.rpc.thrift.TSStatus;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class RatisSchemaRegionStateMachine implements IStateMachine {
+public class SchemaRegionStateMachine extends DataRegionStateMachine {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java:30]:
```suggestion
public class SchemaRegionStateMachine extends BaseStateMachine {
```

File: server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java [hunk 3]
```diff
@@ -39,16 +38,14 @@ public void start() {}
...
+  public TSStatus write(PhysicalPlan plan) {
+    logger.info("Execute write plan in DataRegionStateMachine : {}", plan);
+    return new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode());
   }
 
   @Override
-  public DataSet read(IConsensusRequest request) {
-    logger.info("Execute read plan : {}", request);
+  public DataSet read(PhysicalPlan plan) {
+    logger.info("Execute read plan in DataRegionStateMachine: {}", plan);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java:48]:
```suggestion
    logger.info("Execute read plan in SchemaRegionStateMachine: {}", plan);
```

File: server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java (hunk 1 above)
Comment (OneSizeFitsQuorum) [reply server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java]: Fixed

File: server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java (hunk 3 above)
Comment (OneSizeFitsQuorum) [reply server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java]: Fixed

File: server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java (hunk 2 above)
Comment (OneSizeFitsQuorum) [reply server/src/main/java/org/apache/iotdb/db/consensus/statemachine/SchemaRegionStateMachine.java]: Fixed

**Review by qiaojialin — APPROVED** [review]

**Review by neuyilan — APPROVED** [review]
LGTM

**Pull request merged**